apps/aechannel/src/aesc_fsm.erl

```diff
@@ -1832,8 +1836,59 @@ adjust_ttl(undefined) ->
 adjust_ttl(TTL) when is_integer(TTL), TTL >= 0 ->
     TTL.
 
+-spec curr_hash_and_height() -> {aec_blocks:block_header_hash(),
+                                 aec_blocks:height()}.
+curr_hash_and_height() ->
+    TopHeader = aec_chain:top_header(),
+    {ok, Hash} = aec_headers:hash_header(TopHeader),
+    Height = aec_headers:height(TopHeader), 
+    {Hash, Height}.
+
+-spec curr_height() -> aec_blocks:height().
 curr_height() ->
-    aec_headers:height(aec_chain:top_header()).
+    {_, Height} = curr_hash_and_height(),
+    Height.
+
+pick_hash(#data{block_hash_delta = #bh_delta{ not_newer_than = NNT
+                                            , not_older_than = NOT
+                                            , pick           = PickDelta}}) ->
+    %% Using the boundary of the range is a bit too risky with regard of
+    %% synking. That's why we use an offset
+    Offset = min(NOT, NNT + PickDelta),
+    TopHeader = aec_chain:top_header(),
+    Height = aec_headers:height(TopHeader), 
+    %% use upper limit
+    {ok, Header} = aec_chain:get_key_header_by_height(max(Height - NNT -
+                                                          Offset, 0)),
+    {ok, Hash} = aec_headers:hash_header(Header),
+    Hash.
+
+load_pinned_env(?NOT_SET_BLOCK_HASH) ->
+    {_OnChainEnv, _OnChainTrees} =
+        aetx_env:tx_env_and_trees_from_top(aetx_contract);
+load_pinned_env(BlockHash) ->
+    try aetx_env:tx_env_and_trees_from_hash(aetx_contract, BlockHash) of
```

Comment:
Perhaps these could be cached as well, but `aesc_tx_env_cache.erl` doesn't currently support that.

Reply:
I thought about it but, with forks in mind, it would not be trivial, hence not in the scope of this PR